bin/taste-tester

```diff
@@ -368,6 +375,24 @@ MODES:
     ) do
       options[:skip_post_test_hook] = true
     end
+
+    opts.on(
+      '--use-custom-changeset-hook', 'Find a custom changeset. Works on impact.'
+    ) do
+      options[:use_custom_changeset_hook] = true
+    end
+
+    opts.on(
+      '--use-custom-impact-hook', 'Use customized impact hook. Works on impact.'
```

Comment:
I strongly dislike this, it does not follow the model of the plugins here at all.

Instead, I would model this after the existing plugin infrastructure. I would put the code you want in the common plugin, and let people override it..

that said, what would be *better* would be to find the right hook point sin your common one to make it extensible.

You'll note that taste tester doesnt' provide a way for you to simply write your own upload method, it has a pre_ and post_ upload. For generating the client config, we had `test-remote_client_rb_extra_code`.

I'm happy to help find the right way to factor out the code into reusable parts instead of throwaway parts, but I don't know what specific thigns you want to do internally with your plugin for the phased system so I can't help there... so if you wanna describe that I can probably help.

Reply:
This isn't about phases; it's about being able to override the entire subcommand (`impact`) with custom internal code where applicable.  What's a good pattern for that?

Reply:
@joshuamiller01  why does the internal code need to be different? Without knowing what you want to do internally, it's hard to answer the question. It feels like if the ENTIRE subcommand needs to be different, then (1) the external command isn't that useful OR (2) the external command isn't factored well. The right pattern, were possible, is to have the right hooks throughout the external command so that plugins can shape it's behavior.

Reply:
We can use internal tools to more efficiently determine the affected roles.  Ideally those are open sourced some day, but since they're not in ruby it'll be nontrivial and isn't in scope here.

If the factoring is the problem, them maybe we should have a hook for setting the mechanism which resolves the deps, but the portion which looks at the repository contents and builds the set of affected cookbooks can be standard both internally and open source.   @emroch what do you think?

Reply:
Sure, so having something like TT build a hash object of the source changes and their dependencies, and then hand that too like a `post_process_changes` hook which can than take that and do lookups for like what systems run those cookbooks/roles/whatever.

Then you can have something like a `format_changes` hook, which takes the hash and either returns `nil` if it'd rather the TT format the changes or returns it's own string if it wants to format it itself.

Something like that?

Reply:
Yes, but the calculating of the dependencies is where the internal and external paths diverge.  Knife should work for the majority of cases, even if it is relatively slow.  Our internal tool is both faster and understands our multilayer repository, which Knife cannot do.  I agree that passing the output of the dependency finder to a post_process hook is the way to go, as then each user can configure the output to be useful (hosts, tiers, roles, etc.).